### PR TITLE
Duplicate subsidiary validation added

### PR DESCRIPTION
**Pull request by kaderhashmieviden**

### CODE REVIEW

File: src/EPR.SubsidiaryBulkUpload.Function/EPR.SubsidiaryBulkUpload.Function.csproj [hunk 1]
```diff
@@ -13,6 +13,7 @@
   </PropertyGroup>
   <ItemGroup>
 	  <FrameworkReference Include="Microsoft.AspNetCore.App" />
+	  <None Include="local.settings.json" />
```
Comment (mikeymonster) [inline src/EPR.SubsidiaryBulkUpload.Function/EPR.SubsidiaryBulkUpload.Function.csproj:16]: I don't think you need this - there should already be an item in this .csproj file with local.settings.json

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs [hunk 2]
```diff
@@ -65,39 +66,73 @@ public async Task NotifyErrors(IEnumerable<CompaniesHouseCompany> data, UserRequ
...
 
+    private async Task<List<CompaniesHouseCompany>> FilterDuplicateSubsidiaries(
+        UserRequestModel userRequestModel,
+        (ParentAndSubsidiaries SubsidiaryGroup, OrganisationResponseModel? ParentOrg) subsidiaryGroupAndParentOrg)
+    {
+        var duplicatesGroup = subsidiaryGroupAndParentOrg.SubsidiaryGroup.Subsidiaries.GroupBy(companiesHouseCompany => new
+        {
+            companiesHouseCompany.subsidiary_id,
+            companiesHouseCompany.organisation_name,
+            companiesHouseCompany.companies_house_number,
```
Comment (abidkayanidefra) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs:112]: @kaderhashmieviden  we have discussed this block.
core group id is Companies house number, additional are org name and we have subsidiary id added now. rest is not required. as explained child items already filtered out before reaching this iteration

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs [hunk 3]
```diff
@@ -65,39 +66,73 @@ public async Task NotifyErrors(IEnumerable<CompaniesHouseCompany> data, UserRequ
...
+
+        var subsidiariesToProcess = subsidiaryGroupAndParentOrg.SubsidiaryGroup.Subsidiaries;
+
+        if (duplicatesGroup.Exists(g => g.Count() > 1))
+        {
+            var duplicateItems = subsidiaryGroupAndParentOrg.SubsidiaryGroup.Subsidiaries.Where(company =>
+
+                company.subsidiary_id == duplicatesGroup[0].Key.subsidiary_id &&
+                company.organisation_name == duplicatesGroup[0].Key.organisation_name &&
+                company.companies_house_number == duplicatesGroup[0].Key.companies_house_number &&
```
Comment (abidkayanidefra) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs:124]: this block required to be excluding the duplicates from the remaining rows. duplicates shouldn't be sent for process.

File: src/EPR.SubsidiaryBulkUpload.Function/EPR.SubsidiaryBulkUpload.Function.csproj (hunk 1 above)
Comment (kaderhashmieviden) [reply src/EPR.SubsidiaryBulkUpload.Function/EPR.SubsidiaryBulkUpload.Function.csproj]: Removed

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs (hunk 3 above)
Comment (kaderhashmieviden) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs:124]: The RemoveAt will remove the first record (which we want to preserve) and the line 

`subsidiariesToProcess = subsidiariesToProcess.Except(duplicateItems).ToList();`

will only send the first record only, not the duplicates

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs (hunk 2 above)
Comment (kaderhashmieviden) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/BulkUploadOrchestration.cs:112]: This code identifies the duplicates using all 3 columns in a group. The duplicates are not being filtered before this point (happy to demo on my machine using your example file).

**Review by abidkayanidefra — APPROVED** [review]

**Review by mikeymonster — DISMISSED** [review]

**Review by mikeymonster — DISMISSED** [review]

**Review by MVarkusEviden — APPROVED** [review]

**Review by abidkayanidefra — APPROVED** [review]

**Pull request merged**